src/metabase/lib/schema/expression/conditional.cljc

```diff
@@ -75,26 +75,63 @@
 ;;; believe it or not, a `:case` clause really has the syntax [:case {} [[pred1 expr1] [pred2 expr2] ...]]
 ;;; `:if` is an alias to `:case`
 (doseq [tag [:case :if]]
-  (mbql-clause/define-catn-mbql-clause tag
-    ;; TODO -- we should further constrain this so all of the exprs are of the same type
-    [:pred-expr-pairs [:sequential {:min 1} [:tuple
-                                             {:error/message "Valid [pred expr] pair"}
-                                             #_pred [:ref ::expression/boolean]
-                                             #_expr [:ref ::expression/expression]]]]
-    [:default [:? [:schema [:ref ::expression/expression]]]])
+  (mbql-clause/define-mbql-clause
+    tag
+    [:schema
+     [:and
+      (mbql-clause/catn-clause-schema tag
+                                      [:pred-expr-pairs
+                                       [:sequential {:min 1}
+                                        [:tuple
+                                         {:error/message "Valid [pred expr] pair"}
+                                         #_pred [:ref ::expression/boolean]
+                                         #_expr [:ref ::expression/expression]]]]
+                                      [:default [:? [:schema [:ref ::expression/expression]]]])
+      [:fn
+       ;; Further constrain this so all of the exprs are of compatible type
+       ;; This check isn't perfect: it still allows types that have a common ancestor but aren't compatible.
+       ;; We currently don't have a good notion of type compatibility, so this is just a best-effort heuristic.
+       {:error/message "All clauses should have the same type"}
+       (fn
+         [[_tag _opts pred-expr-pairs default]]
+         (let [exprs (concat
+                      (map second pred-expr-pairs)
+                      (when (some? default)
+                        [default]))
```

Comment:
You could `(cond->> (map ...) (some? default) (cons default))`, since we don't care about the  order.

Reply:
I'll keep it as is for consistency with the `expression/type-of-method` function below, where the order _does_ matter.